python_modules/dagster/dagster/daemon/controller.py

```diff
@@ -85,14 +87,18 @@ def daemon_controller_from_instance(
 
             # Create this in each daemon to generate a workspace per-daemon
             @contextmanager
-            def gen_workspace(_instance):
-                with DaemonWorkspace(grpc_server_registry, workspace_load_target) as workspace:
-                    yield workspace
+            def _context_fn():
+                with DaemonWorkspace(
+                    grpc_server_registry, workspace_load_target
+                ) as workspace, DagsterInstance.from_ref(  # clone instance object so each thread has its own
+                    instance.get_ref()
```

Comment:
get_ref is thread safe right? an alternative would be to pass in the instance_ref to _context_fn